src/charm.py

```diff
@@ -756,8 +831,30 @@ def _on_stop(self, event) -> None:
                 timeout -= 1
                 if timeout < 0:
                     raise Exception(f"{self.unit.name}.on_stop timeout exceeded")
-            logger.debug(f"{self.unit.name} releasing on_stop")
-            self.unit_peer_data["unit_departed"] = ""
+            logger.debug("{self.unit.name} releasing on_stop")
+            self.unit_departed = False
+
+        current_unit_number = unit_number(self.unit)
+        # Raise partition to prevent other units from restarting if an upgrade is in progress.
+        # If an upgrade is not in progress, the leader unit will reset the partition to 0.
+        if k8s_upgrade.partition.get(app_name=self.app.name) < current_unit_number:
+            k8s_upgrade.partition.set(app_name=self.app.name, value=current_unit_number)
+            logger.debug(f"Partition set to {current_unit_number} during stop event")
+        if not self.upgrade._upgrade:
+            logger.debug("Peer relation missing during stop event")
+            return
+
+        # According to the MongoDB documentation, before upgrading the primary, we must ensure a
+        # safe primary re-election.
+        try:
+            if self.unit.name == self.primary:
+                logger.debug("Stepping down current primary, before upgrading service...")
+                self.upgrade.step_down_primary_and_wait_reelection()
+        except FailedToElectNewPrimaryError:
+            logger.error("Failed to reelect primary before upgrading unit.")
+            return
+
+        self.upgrade._upgrade.unit_state = UnitState.RESTARTING
```

Comment:
@carlcsaposs-canonical should review this bit - I am not 100% sure if this is where it should go, but I trust you

QQ: is this meant to handle the entire upgrade workflow? 

other Q: is it necessary to emit a post-upgrade check [like we do in MongoDB VM](https://github.com/canonical/mongodb-operator/blob/9fe3985c8be15d9e7e3d11a14c6b5a3dd6cebb23/src/upgrades/machine_upgrade.py#L191)

Reply:
> QQ: is this meant to handle the entire upgrade workflow?
No, this only handles only the part of the unit that is being refreshed.
This is mostly a safeguard here, based on [mysql k8s router](https://github.com/canonical/mysql-router-k8s-operator/blob/70d6969a66ef1d7d299b3efc3cdbc3e4c39187d4/src/kubernetes_charm.py#L292) `on_stop` handler.
The second part is the check to ensure we're not primary before restarting, and then set the right status.

> other Q: is it necessary to emit a post-upgrade check [like we do in MongoDB VM](https://github.com/canonical/mongodb-operator/blob/9fe3985c8be15d9e7e3d11a14c6b5a3dd6cebb23/src/upgrades/machine_upgrade.py#L191)

Not at that moment. It will happen but after the new pod has started.
I'm following the [upgrades v2 spec](https://docs.google.com/document/d/1tLjknwHudjcHs42nzPVBNkHs98XxAOT2BXGGpP7NyEU/view) here: On all relation events + status, start and pebble ready, we run the `self.upgrade._reconcile_ugprade(event)` which will fire the post upgrade check events.
It cannot be fired before as we don't have updated yet.
Quick explanations:
 * On VM, what happens is `refresh -> (charm code upgraded) -> upgrade-charm event -> …`
 * On k8s, what happens is `refresh -> stop top most unit -> new pod -> upgrade-charm -> start -> pebble ready -> …` so as the unit restarts, we need to split the `upgrade` part in two.

Reply:
> it's important the unit_state is set to restarting or the leader unit could resume the upgrade before the first unit upgrades
I'll double check and add some safeguards but we do have to step down from the replicaset before restarting on MongoDB side. As it's in the unit databag, will it be set eagerly in the databag or does it wait for the event to succeed?

Reply:
It's not unrecoverable but it will leave a downtime if we stop before the unit has stepped down.
Maybe what can be done is to ensure that this code doesn't take more than 25s.
We already have a`Retrying(stop=stop_after_attempt(30), wait=wait_fixed(1), reraise=True)` in the `self.upgrade.step_down_primary_and_wait_reelection()` call so maybe fine tuning this can do it?

Reply:
Would it be possible to change the graceful shutdown timeout you think?
The documentation is quite strict on the fact that we need to elect a new primary before stopping to avoid rollbacks/downtime.
If we can change the graceful shutdown timeout around upgrades, it would solve this issue, if not, I'll see what can be done.

Reply:
@carlcsaposs-canonical After a thorough read of our code + the documentation of mongodb, it shouldn't be a problem since : We leave 10s to elect a new primary and if it fails we have bigger problems.
What would happen if we hadn't returned before the 30s, will the databag still be updated on not?